### DISCUSSION TITLE
WMTS and reprojection

### DISCUSSION THREAD

**Discussion by ShelleyLake** [openlayers/openlayers#16324] [Q&A] [answered]
Dear Team/Community,
I have a request related to the topic of WMTS and reprojection.
I premise that I may have an incorrect expectation, in case maybe someone can give me some guidance.

In short:
- One common Source WMTS (IGN) - `https://data.geopf.fr/wmts?SERVICE=WMTS&VERSION=1.0.0&REQUEST=GetCapabilities`
- View A in "EPSG:3857" (as defined in the WMTS)
- View B in "EPSG:2154" (reprojection)
- Same Zoom level on both Views (21)
- Same center Coordinates on both Views (lng, lat) : [2.33584264266231, 48.8610227385562])

According to the network log:
- View A gets the TileMatrix "21"
- View B gets the TileMatrix "20"

![image](https://github.com/user-attachments/assets/af01a652-1854-426f-ae35-b68270a79d86)

Am I wrong to expect **the same TileMatrix** to be used when **the Zoom is the same**?
Or ... any way to achieve this?

Code here: [https://codesandbox.io/p/sandbox/wmts-and-reprojection-16324-7pgxd5](https://codesandbox.io/p/sandbox/wmts-and-reprojection-16324-7pgxd5)

Any other suggestions are welcome.

Thanks J.

**Comment by mike-000** [comment]
That is expected - in EPSG:2154 one meter in projection units is very close to 1 meter on the ground anywhere in France while in EPSG:3857 one meter in projection units is 1 meter on the ground only at the equator - at 60 degrees latitude is is 0.5 meters on the ground (which is why Greenland looks so much bigger than it really is) and in Paris it is approximately 0.658 meters.  If you add ScaleLine controls to each map you would see the scales are different at the same zoom level.  Reprojection will use the zoom level from the source tiles which best matches the view projection at a particular location - notice how the the zoom level of the OSM tiles changes with latitude in https://openlayers.org/en/latest/examples/reprojection-wgs84.html  See also https://openlayers.org/en/latest/examples/projection-and-scale.html.

**Comment by ShelleyLake** [comment]
@mike-000  Thank you for the explanation, but the problem somehow remains for me.

Even though the TileMatrix  in my example seems to be the same photo scaled to a different resolution, other TileMatrixes (mainly street maps, but the same goes for different zoom levels in Openstreetmap) may show different information.

Layer GEOGRAPHICALGRIDSYSTEMS.PLANIGNV2 - TileMatrix 18
![image](https://github.com/user-attachments/assets/c6aa1b8f-1248-41ec-bedd-344053b57010)

Layer GEOGRAPHICALGRIDSYSTEMS.PLANIGNV2 - TileMatrix 19
![image](https://github.com/user-attachments/assets/faf41c16-b2cf-415d-afb3-3cb39c94e77f)

So if the reprojection algorithm determines, according to a certain formula, that the best TileMatrix for Paris is a different one than expected, well, here I have a problem: different/less information may be shown!

So I try to ask the question in a different way.

Ideally, I would need a way to influence how the reprojection chooses one TileMatrix over another.
(unclear whether this collides with the view/resolution/whatever).

Or maybe a way to find out which TileMatrix is used based on coordinates, resolution and projection (which may be different from the source).
In case the index of the TileMatrix is not the one expected, I can use a higher zoom level. Less elegant, but it might work.

Thanks in advance for any help!

**Reply by mike-000** [reply]
You would need to change the point resolution function of one of the projections so both return the same result

Either set EPSG:2154 to match EPSG:3857
```
const proj2154 = getProjection('EPSG:2154');
proj2154.setGetPointResolution(
  (r, c) => getPointResolution('EPSG:3857', r, transform(c, 'EPSG:2154', 'EPSG:3857'))
);
```
You would need to use `proj2154` as view projection instead of the EPSG string.

Or you could change the built in EPSG:3857 to return projection units
`getProjection('EPSG:3857').setGetPointResolution(r => r);
`

Doing either of these might have undesired side effects, such as incorrect scale lines.

**Reply by jahow** [reply]
What about forcing a different pixel ratio? That would cause different tiles to show up right?

**Reply by mike-000** [reply]
I think point resolutions is the correct way, but to avoid potential side effects override the reprojected source projection with a clone of EPSG:3857 (either defined in proj4 or using addCoordinateTransforms) so the point resolution can be safely changed without any side effects https://stackblitz.com/edit/js-gsvmis?file=index.html,package.json,index.js

**Reply by ShelleyLake** [reply]
Thanks @mike-000, I have tried both your suggestions and they seem to suit me!
If you think that the second method, that overriding the projection of the WMTS, is preferable to the first suggestion, I welcome it!
I just wanted to ask you if, in order to make the code a bit more generic - i.e. without having to go and investigate what the WMTS projection is - it would be possible to define the projection in this way:

```
    const options = optionsFromCapabilities(result, {
      layer: 'GEOGRAPHICALGRIDSYSTEMS.PLANIGNV2',
      matrixSet: 'PM_0_19',
    });

    const source1 = new WMTS(options);

    const projWMTS = getProjection(source1.getProjection()); // <-- !
    projWMTS.setGetPointResolution((r) => r);

    options.projection = projWMTS;
    const source2 = new WMTS(options);
```

[https://codesandbox.io/p/sandbox/wmts-and-reprojection-16324-forked-my68sd](https://codesandbox.io/p/sandbox/wmts-and-reprojection-16324-forked-my68sd)

![image](https://github.com/user-attachments/assets/554f9237-7efd-49ce-88e9-abe95ed0b752)

**Reply by mike-000** [reply, marked as answer]
Since `source1.getProjection()` is EPSG:3857 that would update EPSG:3857 and affect for example a scale line in the EPSG:3857 view.  It would be better to create a new projection using that projection's units and transforms:

    const projWMTS = new Projection({
      code: 'WMTS',
      units: source1.getProjection().getUnits(),
      getPointResolution: (r) => r,
    });
    addCoordinateTransforms(
      projWMTS,
      'EPSG:2154',
      (c) => transform(c, source1.getProjection(), 'EPSG:2154'),
      (c) => transform(c, 'EPSG:2154', source1.getProjection())
    );

https://stackblitz.com/edit/js-k4b5g7?file=index.html,package.json,index.js

**Reply by ShelleyLake** [reply]
Thank you very much!
